modules/http4s/src/smithy4s/http4s/internals/SimpleRestJsonCodecs.scala

```diff
@@ -73,38 +79,73 @@ private[http4s] class SimpleRestJsonCodecs(
       .withContentType(`Content-Type`(mediaType"application/json"))
   }
 
+  private type EncoderFromHints[F[_], A] = Hints => ResponseEncoder[F, A]
+
   def makeServerCodecs[F[_]: Concurrent]: UnaryServerCodecs.Make[F] = {
     val messageDecoderCompiler =
       RequestDecoder.restSchemaCompiler[F](
         Metadata.Decoder,
         entityDecoders[F]
       )
+    val restSchemaCompiler = ResponseEncoder.restSchemaCompiler[F](
+      Metadata.Encoder,
+      entityEncoders[F]
+    )
+
     val responseEncoderCompiler = {
-      val restSchemaCompiler = ResponseEncoder.restSchemaCompiler[F](
-        Metadata.Encoder,
-        entityEncoders[F]
-      )
-      new CachedSchemaCompiler[ResponseEncoder[F, *]] {
+      new CachedSchemaCompiler[EncoderFromHints[F, *]] {
         type Cache = restSchemaCompiler.Cache
         def createCache(): Cache = restSchemaCompiler.createCache()
         def fromSchema[A](schema: Schema[A]) = fromSchema(schema, createCache())
 
-        def fromSchema[A](schema: Schema[A], cache: Cache) = if (schema.isUnit) {
-          restSchemaCompiler.fromSchema(schema, cache)
-        } else {
-          restSchemaCompiler
-            .fromSchema(schema, cache)
-            .andThen(addEmptyJsonToResponse(_))
-        }
+        def fromSchema[A](schema: Schema[A], cache: Cache) = endpointHints =>
```

Comment:
That's better, but an arguably better approach would be to modify Schema before it's passed to this EndpointCompiler by forwarding the hint to the output schema, instead of changing the type the compiler works against : 

``` 
output.fromSchema(endpoint.output.addHints(endpoint.hints), responseEncoderCache)
```

Reply:
is there a possibility that the hints we would be adding from the endpoint affect the compiler behaviour?

I don't think the `smithy.api.Http` has a chance of doing that but maybe other hints could?

can't think of such hints at the moment but eh

Reply:
@daddykotex I don't think it's much of concern here, but if you're actually worried about it, it's totally possible to create private hints directly in Scala and use them it ways that would have pretty much zero chance of interfering with existing interpreter behaviour.